jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java

```diff
@@ -828,6 +849,150 @@ MethodHandle insertArguments(MethodHandle equivalent, MethodHandle unboxingHandl
+			long bsmArgs = bsmData + BSM_ARGUMENTS_OFFSET;
+			MethodHandle bsm = getCPMethodHandleAt(internalRamClass, bsmIndex);
+			if (null == bsm) {
+				/*[MSG "K05cd", "unable to resolve 'bootstrap_method_ref' in '{0}' at index {1}"]*/
+				throw new NullPointerException(Msg.getString("K05cd", classObject.toString(), bsmIndex)); //$NON-NLS-1$
+			}
+			Object[] staticArgs = new Object[BSM_OPTIONAL_ARGUMENTS_START_INDEX + bsmArgCount];
+			/* Mandatory arguments */
+			staticArgs[BSM_LOOKUP_ARGUMENT_INDEX] = new MethodHandles.Lookup(classObject, false);
+			staticArgs[BSM_NAME_ARGUMENT_INDEX] = name;
+			staticArgs[BSM_TYPE_ARGUMENT_INDEX] = typeClass;
+		
+			/* Static optional arguments */
+			/* internalRamClass is not a j.l.Class object but the ConstantPool natives know how to
+			 * get the internal constantPool from the j9class
+			 */
+			ConstantPool cp = access.getConstantPool(internalRamClass);
+
+			/* Check if we need to treat the last parameter specially when handling primitives.
+			 * The type of the varargs array will determine how primitive ints from the constantpool
+			 * get boxed: {Boolean, Byte, Short, Character or Integer}.
+			 */ 
+			boolean treatLastArgAsVarargs = bsm.isVarargsCollector();
+			Class<?> varargsComponentType = bsm.type.lastParameterType().getComponentType();
+			int bsmTypeArgCount = bsm.type.parameterCount();
+
+			for (int i = 0; i < bsmArgCount; i++) {
+				int staticArgIndex = BSM_OPTIONAL_ARGUMENTS_START_INDEX + i;
+				short index = UNSAFE.getShort(bsmArgs + (i * BSM_ARGUMENT_SIZE));
+				int cpType = getCPTypeAt(internalRamClass, index);
+				Object cpEntry = null;
+				switch (cpType) {
+				case 1:
+					cpEntry = cp.getClassAt(index);
+					if (cpEntry == null) {
+						throw throwNoClassDefFoundError(classObject, index);
+					}
+					break;
+				case 2:
+					cpEntry = cp.getStringAt(index);
+					break;
+				case 3: {
+					int cpValue = cp.getIntAt(index);
+					Class<?> argClass;
+					if (treatLastArgAsVarargs && (staticArgIndex >= (bsmTypeArgCount - 1))) {
+						argClass = varargsComponentType;
+					} else {
+						argClass = bsm.type.parameterType(staticArgIndex);
+					}
+					if (argClass == Short.TYPE) {
+						cpEntry = (short) cpValue;
+					} else if (argClass == Boolean.TYPE) {
+						cpEntry = cpValue == 0 ? Boolean.FALSE : Boolean.TRUE;
+					} else if (argClass == Byte.TYPE) {
+						cpEntry = (byte) cpValue;
+					} else if (argClass == Character.TYPE) {
+						cpEntry = (char) cpValue;
+					} else {
+						cpEntry = cpValue;
+					}
+					break;
+				}
+				case 4:
+					cpEntry = cp.getFloatAt(index);
+					break;
+				case 5:
+					cpEntry = cp.getLongAt(index);
+					break;
+				case 6:
+					cpEntry = cp.getDoubleAt(index);
+					break;
+				case 13:
+					cpEntry = getCPMethodTypeAt(internalRamClass, index);
+					break;
+				case 14:
+					cpEntry = getCPMethodHandleAt(internalRamClass, index);
+					break;
+				default:
```

Comment:
also need a `case 17:` for constant dynamic. Right now this will be a 
```
//TODO
break;
```

until https://github.com/eclipse/openj9/pull/1649 is complete

`resolveInvokeDynamic` will have to be updated as well